### PR TITLE
Fuzz test location-related parser options

### PR DESCRIPTION
**Pull request by tolmasky**
| Q                        | A
| ------------------------ | ---
| Fixed Issues?            | 
| Patch: Bug Fix?          | Yes
| Major: Breaking Change?  | No
| Minor: New Feature?      | No
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

This PR completes the work started in https://github.com/babel/babel/pull/14130 by adding the necessary foundations to do fuzz testing in our tests, and implements the first instance of such fuzz testing with `startLine` and `startColumn`. Basically, it takes all the existing tests, and runs them all with several different combinations of `startLine` and `startColumn` (including having them absent in the `options` object, etc.). This increases the number of total tests we run from `22566` to `88344`.

1. Despite effectively quadrupling the number of tests, they actually complete in the same amount of time or faster. For example, the Node.js 14 tests complete in 109s vs. 119s in CI. This is due to a number of performance improvements in the way we compare ASTs, etc. There's way more performance stuff I'd like to do, but I tried to limit this PR to as small of a footprint as possible. By changing `getFixtures` and the actual test format, it should be possible to dramatically improve the performance of tests further (especially when running only a subset with TEST_GREP, which currently still pays a huge price for all the tests it doesn't run).

2. The sole exception to this is node 8, where tests were taking an extremely long time after this change. After spending way too much time trying to figure this out, I discovered that Jest 23 on node 8 just has a threshold number of tests where afterward it simply takes forever (40 min+) no matter what, even if the tests are completely empty. As such, I've simply turned off fuzz testing in Node 8, given that support for node 8 is going to be dropped in Babel 8 anyways. Interestingly, Node 6 does *not* have this problem.

3. You can turn fuzz testing off by setting the `FUZZ` environment variable to false: `FUZZ=false`. There should really be no reason to though, as the tests take the same amount of time to complete.

4. This PR also fixes showing code frames, which was broken I believe in this [commit](https://github.com/babel/babel/pull/13966) due to calling `error.toString()`, which does  not include the stack. The behavior is actually improved, as you'll now see code frames even in the recovered errors:

Screen Shot 2022-01-25 at 4 17 31 PM

5. I renamed "runFixtureTests.js" to "run-fixture-tests.js" to avoid case sensitive filesystem issues. Happy to change this back if it is not desired.

6. There's a `Difference` class that calculates differences between ASTs in a way that's easy to manipulate after-the-fact:

    ```javascript
    const difference = new Difference(adjust, expect, actual);

    // difference.path === [`ast`, `body`, 0, `expression`, `value`]
    // difference.reason === { type: "value", expected: 5, actual: 6 }
    // difference.message = "5 != 6 in .ast.body[0].expression.value"
    ```

    As you can see, the error message now uses a "JavaScript compliant" accessor path (".ast.body[0].expression.value"), instead of the "filesystem-style" path it used to present (something like "/ast/body/0/expression/value"). This makes it easy to copy-paste the path into a REPL if you are trying to debug the problem. It also fixes a bug with the regex replacement stuff that was done with string manipulation in the old system. It uses `isIdentifierName` internally, so it should always generate a JavaScript-friendly path.

    You'll notice there is an `adjust` function that is passed to `Difference`. This is one of the things that allows us to make fuzz testing quick. Instead of either generating a different AST for each expected parse with different options, or even simply "modifying" the expected AST for each fuzz option, the `adjust` function travels along the comparison calls (much like the `serialize` function in `JSON.stringify`), and is given a chance to modify the value that is compared. So, for example, if the fuzzed options were "startLine = 10, startColumn = 20", the generated `adjust` function does something along these lines:

    ```javascript
    function adjust(value, key, original) {
      if (key === "line") {
         return value - 1 + startLine; // 1 -> 20, 2 -> 21, etc.
      }
      // Only adjust the column for "line 1"
      if (key === "column" && original.line === 1) {
         return value + startColumn;
      }
      return value;
    }
    ```

    It's a bit more complicated since the function is auto-generated of course, and because it also has to account for if the original test *itself* included a startLine, but this is the basic idea. You can see that this could fairly trivially be expanded to automatically testing `errorRecovery` being `true` and `false` (explained in more detail below).

7. `FixtureError` has been expanded to give some additional semantic meaning. There's now `FixtureError.DifferentAST`, `FixtureError.UnexpectedError`, `FixtureError.UnexpectedRecovery`, etc. This generally makes it easier to keep track of what's going on and generate nice error messages, instead of repeatedly throwing and catching and modifying message strings. The correct `FixtureError` is obtained by passing the `Difference`:

    ```javascript
    const error = FixtureError.fromDifference(difference, actual);
    ```

8. There is no more need for `runFixtureTestsWithoutExactASTMatch`, as the `onlyCompareErrors` option has been pushed up to the `runFixtureTests` API, so it can simply be called with `true` as the last argument to get this behavior.

9. As I mentioned above, I specifically stayed away from changing file formats, but I think one of the next steps is to properly serialize errors. We currently don't actually really test whether `startLine` and `startColumn` affect errors, since we only serialize their error messages (and in subtly different ways depending on whether the error was recovered from or not). We do however test this when the line and column number happen to be in the error message itself.

With this foundation in place, my plan is to slowly incorporate more of the options into our fuzz testing, to get more bang for our buck from our existing tests, and to additionally discover more edge cases. For example, the one I'd like to do next is to automatically try each test with errorRecovery set to both `true` and `false` (and missing). Given that tests are already generated with errorRecovery set to `true`, when we run it with `false`, we simply have to test that the error thrown is equal to the first error in the errors array. This fuzz test would have for example caught this bug: https://github.com/babel/babel/issues/14146

Generally speaking, the "ideal" would be to eventually handle *all* options through this system, such that the total number of effective tests is `[base_tests] * 2 ^ (# of boolean options) * (# of multi-state-options ...)`. Through performance work on the tests like the above, this should have negligible effects on our runtime, but potentially offer a great benefit in discovering bugs when running babel with complicated combinations of options without requiring a ton of manual labor to keep adding such tests.

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/14201"></a>

### CODE REVIEW

**Comment by tolmasky** [comment]
I forgot to include one minor point: I made a small format change in that the 2 (!) tests that use the babel serialization key to encode either `RegExp`s or `bigint`s  in their AST JSON files now have the file extension ".extended.json" instead of ".json". This just helps a ton by allowing me to entirely avoid passing a `reviver` function to `JSON.parse` for the thousands of other tests that *don't* do any special encoding in their JSON serialization. The test generator is smart and will automatically give the file the correct extensions, so it shouldn't add any work for anyone.

File: packages/babel-parser/test/helpers/polyfill.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
+// We run these tests as far back as Node 6, so we need these there.
```
Comment (nicolo-ribaudo) [inline packages/babel-parser/test/helpers/polyfill.js:1]: Can you just add a `// TODO(Babel 8): remove this file` comment? So that we don't forget.

File: packages/babel-parser/test/helpers/to-contextual-syntax-error.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
+import { codeFrameColumns } from "@babel/code-frame";
+import { runInThisContext } from "vm";
+
+export default function toContextualSyntaxError(
```
Comment (nicolo-ribaudo) [inline packages/babel-parser/test/helpers/to-contextual-syntax-error.js:4]: Q: What do we need this for?

File: packages/babel-parser/test/helpers/to-fuzzed-options.js [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    (object, [key, descriptor]) => (
+      !descriptor || descriptor.delete
+        ? delete object[key]
+        : Object.defineProperty(object, key, descriptor),
+      object
+    ),
+    clone(object),
+  );
+
+const toAdjust = adjustments =>
```
Comment (nicolo-ribaudo) [inline packages/babel-parser/test/helpers/to-fuzzed-options.js:18]: Nit: `toAdjustFunction`, it makes it a bit clear to understand what this should do.

File: packages/babel-parser/test/helpers/to-fuzzed-options.js [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+      options,
+    ])
+    .map(([adjust, options]) => [
+      adjust &&
+        toAdjust({
+          ...adjust,
+          threw: (adjust, error) =>
+            error && toAdjustedSyntaxError(adjust, error),
+          errors: (adjust, errors) =>
+            errors && errors.map(error => toAdjustedSyntaxError(adjust, error)),
```
Comment (nicolo-ribaudo) [inline packages/babel-parser/test/helpers/to-fuzzed-options.js:107]: Can these just be moved to the previous `.map`, where we also define `line:` and `column:`?

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -183,12 +183,21 @@ jobs:
         with:
           node-version: ${{ matrix.node-version }}
       - name: Test on node.js ${{ matrix.node-version }}
+        if: matrix.node-version == '8'
```
Comment (nicolo-ribaudo) [inline .github/workflows/ci.yml:186]: Let's also skip them on Node.js 6 and 10: they are painfully slow, and we don't get much benefits from also running fuzz tests there :sweat_smile: 

If/when we'll add more combinations, if it becomes too slow then we can also decide to only run fuzz tests on a single Node.js version.

**Review by nicolo-ribaudo — COMMENTED** [review]
Thanks, this looks really good!

File: packages/babel-parser/test/helpers/difference.js [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+  // easy.
+  if (Object.is(expected, actual)) {
+    return false;
+  }
+
+  const typeExpected = toType(expected);
+  const typeActual = toType(actual);
+
+  if (typeExpected !== typeActual) {
+    return { discrepancy: "value", expected, actual };
```
[the remark this answers is not in the document]
Comment (tolmasky) [reply packages/babel-parser/test/helpers/difference.js:50]: I'll get rid of the "type" handler in `toExplanationString`, I used to have this be `type` but it to be a less useful error message ("expected number but got string", vs. `5 !== "hello"`), so I changed it to this and forgot to get rid of that handler.

File: packages/babel-parser/test/helpers/to-fuzzed-options.js (hunk 4 above)
Comment (tolmasky) [reply packages/babel-parser/test/helpers/to-fuzzed-options.js:107]: The result of the first map lets us know whether we need the second one. If it worked out to there needing to be no adjustment for line and column, then `adjust` will be false and we also won't need to do these.

File: .github/workflows/ci.yml (hunk 5 above)
Comment (tolmasky) [reply .github/workflows/ci.yml:186]: Is there a way to set `TEST_FUZZ` in an earlier rule so that it's true for 12/14/16 and false for 6/8/10?

File: packages/babel-parser/test/helpers/polyfill.js (hunk 1 above)
Comment (tolmasky) [reply packages/babel-parser/test/helpers/polyfill.js:1]: Sure thing.

File: packages/babel-parser/test/helpers/to-contextual-syntax-error.js (hunk 2 above)
Comment (tolmasky) [reply packages/babel-parser/test/helpers/to-contextual-syntax-error.js:4]: So, what we actually wish we had was [FireFox's `Error.fileName`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/fileName), but unfortunately v8 doesn't support this property. So we use `vm.runInContext` to make it think it's coming from this file. This would on its own though make it appear at the bottom of the trace, not the top, because we're really  doing something weird here combining two ideas of stack trace. So instead of fiddling with that, we have the "Contextual Error", which pretends to exist in "filename", and that has as its cause the Babel Error, that way we get the "at filename" followed by the Babel stacktrace for free from node's inspect, vs. having to do string manipulation (which is more fragile and starts to break with more complicated messages).

Comment (nicolo-ribaudo) [reply packages/babel-parser/test/helpers/to-contextual-syntax-error.js:4]: Rather than string manipulation we could have used `Error.captureStackTrace`, but this approach is good too. However, please add a comment explaining this directly in the source code :pray:

File: .github/workflows/ci.yml (hunk 5 above)
Comment (nicolo-ribaudo) [reply .github/workflows/ci.yml:186]: Maybe this, but I didn't try:
```yaml
      - name: Test on node.js ${{ matrix.node-version }}
        # Hack: --color has supports-color@5 returned true for GitHub CI
        # Remove once `chalk` is bumped to 4.0.

        # Todo(Babel 8): Jest execution path is hardcoded because Yarn 2 does not support node 6
        run: |
          node --max-old-space-size=4096 ./node_modules/.bin/jest --ci --color
        with:
          BABEL_ENV: test
          TEST_FUZZ: ${{ matrix.node-version >= 12 }}
```

File: packages/babel-parser/test/helpers/to-contextual-syntax-error.js (hunk 2 above)
Comment (tolmasky) [reply packages/babel-parser/test/helpers/to-contextual-syntax-error.js:4]: Unfortunately `captureStackTrace` doesn't work for this (unless I'm missing part of the API). The goal is to convince the constructor that it's happening inside of a file that it's not actually happening in (`babel-parser/test/.../input.js`), and all I think `captureStackTrace` can do is remove the last function on the stack from showing up (a different problem). Hence the necessary roundtrip through vm to convince node we're "in" that, kind of how node does it themselves: https://github.com/nodejs/node/blob/e46c680bf2b211bbd52cf959ca17ee98c7f657f5/test/message/vm_display_syntax_error.js#L29

Comment (tolmasky) [reply packages/babel-parser/test/helpers/to-contextual-syntax-error.js:4]: And yeah, I'll put a comment in there.

File: packages/babel-parser/test/helpers/difference.js [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+    ? `Array of wrong size, expected length of ${expected}, but got ${actual}`
+    : discrepancy === "unexpected-key"
+    ? `Did not expect a property ${toValueString(key)}`
+    : discrepancy === "missing-key"
+    ? `${toType(actual)} is missing property ${toValueString(key)}`
+    : discrepancy === "type"
+    ? `Wrong type "${expected}" != "${actual}"`
+    : `${toValueString(expected)} != ${toValueString(actual)}`;
+
+const isInt = key => parseInt(key, 10) + "" === key;
```
[the remark this answers is not in the document]
Comment (tolmasky) [reply packages/babel-parser/test/helpers/difference.js]: Done: https://github.com/babel/babel/pull/14201/commits/ffb642e15e3163f14937783ce19a083e58fc74e9

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -183,12 +183,21 @@ jobs:
...
         # Hack: --color has supports-color@5 returned true for GitHub CI
         # Remove once `chalk` is bumped to 4.0.
 
         # Todo(Babel 8): Jest execution path is hardcoded because Yarn 2 does not support node 6
         run: |
-          BABEL_ENV=test node ./node_modules/.bin/jest --ci --color
+          BABEL_ENV=test FUZZ=false node --max-old-space-size=4096 ./node_modules/.bin/jest --ci --color
```
[the remark this answers is not in the document]
Comment (tolmasky) [reply .github/workflows/ci.yml]: Done: https://github.com/babel/babel/pull/14201/commits/ffb642e15e3163f14937783ce19a083e58fc74e9

File: packages/babel-parser/test/helpers/polyfill.js (hunk 1 above)
Comment (tolmasky) [reply packages/babel-parser/test/helpers/polyfill.js:2]: Done: https://github.com/babel/babel/pull/14201/commits/ffb642e15e3163f14937783ce19a083e58fc74e9

File: .github/workflows/ci.yml (hunk 5 above)
Comment (tolmasky) [reply .github/workflows/ci.yml]: Done: https://github.com/babel/babel/pull/14201/commits/1a24870b4b7b968cbb71239eeeb7cd19ff401fde

File: packages/babel-parser/test/helpers/difference.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (tolmasky) [reply packages/babel-parser/test/helpers/difference.js:51]: Done: https://github.com/babel/babel/pull/14201/commits/ffb642e15e3163f14937783ce19a083e58fc74e9

File: packages/babel-parser/test/helpers/to-fuzzed-options.js (hunk 3 above)
Comment (tolmasky) [reply packages/babel-parser/test/helpers/to-fuzzed-options.js]: Done: https://github.com/babel/babel/pull/14201/commits/5e760114b5f1650507e12ecec7c605ec74abf80b

File: packages/babel-parser/test/helpers/difference.js [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+      );
+
+const toValueString = (value, type = toType(value)) =>
+  type === "string"
+    ? JSON.stringify(value)
+    : type === "symbol"
+    ? value.toString()
+    : type === "bigint"
+    ? `${value}n`
+    : type !== "number" || 1 / value !== -Infinity
```
[the remark this answers is not in the document]
Comment (tolmasky) [reply packages/babel-parser/test/helpers/difference.js]: Done: https://github.com/babel/babel/pull/14201/commits/5e760114b5f1650507e12ecec7c605ec74abf80b

File: .github/workflows/ci.yml (hunk 8 above)
[the remark this answers is not in the document]
Comment (tolmasky) [reply .github/workflows/ci.yml]: With regard to the memory stuff, I had originally put it in there because in a very early version it was running out of memory, but no longer. However, it does help with performance generally, since the older versions of node have a bug where they can't automatically address more than 2GB (as opposed to older versions which can without any special flag), and the fact that the tests are written in a way where before any single test is run node is basically taken to the very threshold of its memory limit, because it starts by reading every single test file, before even deciding whether to skip tests, etc. So basically it puts the entirety of all the tests into memory and keeps them there throughout most of the run. Hence any additional memory it can get helps tremendously (I think it shaves a minute of Node 6 runtime -- regardless of fuzz testing or not). This is something I'm planning to attack in my next PR: only read the files when its actually time to run the tests, such that with TEST_GREP thousands of files don't get read at all.

File: packages/babel-parser/test/helpers/fixture-error.js [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+      UnexpectedRecovery: ({ expected }, errors) =>
+        `Expected unrecoverable error:\n\n    ${expected}\n\n` +
+        `But instead parsing recovered from ${errors.length} errors.`,
+    },
+    ([name, toMessage]) => [
+      name,
+      named(
+        name,
+        class extends FixtureError {
+          constructor(difference, cause) {
```
[the remark this answers is not in the document]
Comment (tolmasky) [reply packages/babel-parser/test/helpers/fixture-error.js:61]: Sure, just as an aside, unrelated to this, it's weird that Node seems to have a bug where if you do `new Error("blah", { cause })`, the cause *won't* show up in the stack trace. However, if you do `x = new Error("blah"); x.cause = ...`, then the cause *will* show up in the stack trace. This is because passing cause to the constructor form makes it so cause will not be enumerable, and thus `inspect` seems to miss it. So as of node 16, the only way to see `cause` is to *not* use the standard constructor form.

File: packages/babel-parser/test/helpers/to-contextual-syntax-error.js (hunk 2 above)
Comment (tolmasky) [reply packages/babel-parser/test/helpers/to-contextual-syntax-error.js:4]: Done: https://github.com/babel/babel/pull/14201/commits/f6aeb0651250a51ec6b0838a6525a8364e96b82c

File: packages/babel-parser/test/helpers/fixture-error.js [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+            if (cause) {
+              this.cause = isArray(cause)
+                ? cause.map(cause => cause.context || cause)
+                : cause;
+            }
+          }
+
+          // Don't show the stack of FixtureErrors, it's irrelevant.
+          // Instead, show the cause, if present.
+          [inspect.custom](depth, options) {
```
[the remark this answers is not in the document]
Comment (tolmasky) [reply packages/babel-parser/test/helpers/fixture-error.js]: Done: https://github.com/babel/babel/pull/14201/commits/f6aeb0651250a51ec6b0838a6525a8364e96b82c

File: packages/babel-parser/test/helpers/fixture-error.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (tolmasky) [reply packages/babel-parser/test/helpers/fixture-error.js]: Done: https://github.com/babel/babel/pull/14201/commits/f6aeb0651250a51ec6b0838a6525a8364e96b82c

**Comment by tolmasky** [comment]
I *think* that addresses all the comments, but please let me know if I missed any.

File: packages/babel-parser/test/helpers/fixture-error.js [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+export default class FixtureError extends Error {
+  constructor(difference, { cause } = {}) {
+    super();
+
+    // Sigh, still have to manually set the name unfortunately...
+    named(this.constructor.name, this);
+
+    this.difference = difference;
+
+    // Set cause ourselves, since node < 17 has a bug where it won't show it
```
Comment (nicolo-ribaudo) [inline packages/babel-parser/test/helpers/fixture-error.js:23]: It's not a bug, it's just that `Error#cause` reached stage 4 recently :stuck_out_tongue:

**Review by nicolo-ribaudo — APPROVED** [review]
Thank you!

File: packages/babel-parser/test/helpers/serialization.js [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+// This is because only estree actually includes the BigInt value in the Literal
+// node. If the JS environemnt doesn't support bigint, then estree will just
+// use null for the value. We also happen to just throw the AST information away
+// with estree tests, so in the event that we're running on an older version of
+// Node that doesn't support bigint, it is safe to deserialize to null.
+const toBigInt = global.BigInt || (() => null);
+
+const SerializationKey = "$$ babel internal serialized type";
+
+/* eslint-disable no-confusing-arrow */
```
Comment (JLHwung) [inline packages/babel-parser/test/helpers/serialization.js:23]: Ditto

**Review by JLHwung — COMMENTED** [review]
Will do a more thorough review later.

File: packages/babel-parser/test/helpers/to-fuzzed-options.js [hunk 14]
```diff
@@ -0,0 +1,111 @@
+/* eslint-disable no-confusing-arrow */
```
[the remark this answers is not in the document]
Comment (tolmasky) [reply packages/babel-parser/test/helpers/to-fuzzed-options.js:1]: Sure, is there a workaround other than "don't use arrow functions or don't use ternary operators"? I had turned it on since the eslint docs tell you that this warning wants you to wrap it in parenthesis to resolve the "ambiguity"... only to then complain about un-needed parenthesis.

File: packages/babel-parser/test/helpers/run-fixture-tests.js [hunk 15]
```diff
@@ -0,0 +1,177 @@
...
+      testFn(test.title, () => runParseTest(parse, test, onlyCompareErrors)),
+    );
+}
+
+const toJustErrors = result => ({
+  threw: result.threw,
+  ast: result.ast && { errors: result.ast.errors },
+});
+
+function runParseTest(parse, test, onlyCompareErrors) {
```
[the remark this answers is not in the document]
Comment (tolmasky) [reply packages/babel-parser/test/helpers/run-fixture-tests.js:83]: Done: https://github.com/babel/babel/pull/14201/commits/caa9a8ceb268151389e39d213c24478a3c7a02fc

File: packages/babel-parser/test/helpers/run-fixture-tests.js [hunk 16]
```diff
@@ -0,0 +1,177 @@
...
+  );
+
+  // No differences means we passed and there's nothing left to do.
+  if (difference === Difference.None) return;
+
+  const error = FixtureError.fromDifference(difference, actual);
+
+  // If we're not overwriting the current values with whatever we get this time
+  // around, then we have a legitimate error that we need to report.
+  if (!OVERWRITE) throw error;
```
[the remark this answers is not in the document]
Comment (tolmasky) [reply packages/babel-parser/test/helpers/run-fixture-tests.js]: Done: https://github.com/babel/babel/pull/14201/commits/caa9a8ceb268151389e39d213c24478a3c7a02fc

File: packages/babel-parser/test/helpers/to-fuzzed-options.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (tolmasky) [reply packages/babel-parser/test/helpers/to-fuzzed-options.js:1]: Sorry, to clarify, when I fix the arrow issue, then *prettier* complains about the disambiguating parenthesis. So it's a conflict between these two almost, but not quite, overlapping tools want.

[the remark this answers is not in the document]
Comment (tolmasky) [reply packages/babel-parser/test/helpers/to-fuzzed-options.js:1]: OK, so it looks like I'll just have to change like 20 arrows function to non-arrow functions, since what would be the result from autofixing this (`--fix`) is in conflict with the separate Prettier run. I'll do this if we feel strongly about it, but I'd perhaps like to make an appeal for instead turning this setting off (or being OK with having this comment disabling it here):

1. It's actually off by default in `eslint:recommended`, because as I'll describe below, it's kind of a stretch of an "ambiguity".
2. As described above, it's normally a simple fix (just wrapping in parenthesis), but in our case leads to bouncing back and forth between conflicting style enforcers.
3. The actual case being protected against being "confusing" is IMO somewhat... strange. The idea is to protect against the possibility that the author may have meant "<=" instead of "=>". So, for example, in my function below:

Screen Shot 2022-01-26 at 6 59 47 PM

The idea is that perhaps I *actually* meant:

```javascript
const toExplanationString = ({ discrepancy, expected, actual, key }) <= discrepancy === "length"
```

In other words, the implication is that it's hard to tell whether I *meant* what I typed, an arrow function, or whether what I actually meant was to check whether a number-coerced newly created object (which would be `NaN`) is smaller than or equal to... the boolean result of a triple equals. Seems... unlikely. Especially as we move to typescript across the board.

Anyways, don't meant to be difficult, I'll go and convert them all to non-arrow functions (which unfortunately changes quite a few one-liners to 5-line noisier functions due the indented return then forces another wrap, etc.), but just wanted to throw it out there considering the weird conflict with Prettier, and on the off chance that this was not some super-important linting option but perhaps simply copied over form some other project or something.

[the remark this answers is not in the document]
Comment (JLHwung) [reply packages/babel-parser/test/helpers/to-fuzzed-options.js:1]: Good point! I believe `no-confusing-arrow` rule should allow such cases, at least behind an option. I have open https://github.com/eslint/eslint/issues/15548

Before ESLint team decide, I am good to leave it as-is.

File: packages/babel-parser/test/helpers/run-fixture-tests.js [hunk 17]
```diff
@@ -7,9 +7,23 @@ import toFuzzedOptions from "./to-fuzzed-options.js";
...
 import toContextualSyntaxError from "./to-contextual-syntax-error.js";
 
-const { OVERWRITE } = process.env;
+const { CI, OVERWRITE } = process.env;
 const { stringify, parse: JSONParse } = JSON;
 
+/**
+ * run parser on given tests
+ *
+ * @param {Test} A {@link packages/babel-helper-fixtures/src/index.js Test}
```
Comment (JLHwung) [inline packages/babel-parser/test/helpers/run-fixture-tests.js:16]: The param should be a `string`.

File: packages/babel-parser/test/helpers/to-fuzzed-options.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (tolmasky) [reply packages/babel-parser/test/helpers/to-fuzzed-options.js:1]: Great, thanks!

File: packages/babel-parser/test/helpers/run-fixture-tests.js (hunk 17 above)
Comment (tolmasky) [reply packages/babel-parser/test/helpers/run-fixture-tests.js]: Oops! Fixed, and updated explanation too: https://github.com/tolmasky/babel/commit/9b87eeaeb5ba81d228c92e8a34f236f777d96226

File: packages/babel-parser/test/helpers/run-fixture-tests.js [hunk 18]
```diff
@@ -0,0 +1,189 @@
...
+
+  // We want to throw away the contents of `throws` here.
+  // eslint-disable-next-line no-unused-vars
+  const { throws: _, ...oldOptions } = readJSON(optionsLocation);
+  const newOptions = { ...oldOptions, ...(throws && { throws }) };
+
+  // Store (or overwrite) the options file if there's anything to record,
+  // otherwise remove it.
+  if (Object.keys(newOptions).length > 0) {
+    writeFileSync(optionsLocation, stringify(newOptions, null, 2), "utf-8");
```
Comment (JLHwung) [inline packages/babel-parser/test/helpers/run-fixture-tests.js:128]: To be consistent with `.editorconfig`: https://github.com/babel/babel/blob/8ccd1219f830784ead6178180f87fd23e0bf9bd9/.editorconfig#L4-L7


we should insert a trailing `\n`.

File: packages/babel-parser/test/helpers/run-fixture-tests.js [hunk 19]
```diff
@@ -0,0 +1,189 @@
...
+  }
+
+  // When only comparing errors, we don't want to overwrite the AST JSON because
+  // it belongs to a different test.
+  if (onlyCompareErrors) return;
+
+  const normalLocation = join(testLocation, "output.json");
+  const extendedLocation = join(testLocation, "output.extended.json");
+
+  const [extended, serialized] = actual.ast ? serialize(actual.ast) : [];
```
Comment (JLHwung) [inline packages/babel-parser/test/helpers/run-fixture-tests.js:140]: ditto.

File: packages/babel-parser/test/helpers/run-fixture-tests.js [hunk 20]
```diff
@@ -0,0 +1,189 @@
...
+  const optionsLocation = join(testLocation, "options.json");
+
+  // We want to throw away the contents of `throws` here.
+  // eslint-disable-next-line no-unused-vars
+  const { throws: _, ...oldOptions } = readJSON(optionsLocation);
+  const newOptions = { ...oldOptions, ...(throws && { throws }) };
+
+  // Store (or overwrite) the options file if there's anything to record,
+  // otherwise remove it.
+  if (Object.keys(newOptions).length > 0) {
```
Comment (JLHwung) [inline packages/babel-parser/test/helpers/run-fixture-tests.js:127]: I don't think we need to update `options.json` if `options.json` does not contains `throws` at all, i.e.

https://github.com/babel/babel/blob/8ccd1219f830784ead6178180f87fd23e0bf9bd9/packages/babel-parser/test/fixtures/estree/options.json#L1-L3

I notice this issue because prettier and `JSON.stringify` have different output here.

File: packages/babel-parser/test/helpers/run-fixture-tests.js [hunk 21]
```diff
@@ -0,0 +1,189 @@
...
+    writeFileSync(optionsLocation, stringify(newOptions, null, 2), "utf-8");
+  } else if (existsSync(optionsLocation)) {
```
Comment (JLHwung) [inline packages/babel-parser/test/helpers/run-fixture-tests.js:129]: Checking file existence before deletion is subject to race conditions. We can directly call `unlinkSync` but mute the `ENOENT` errors.

File: packages/babel-parser/test/helpers/to-fuzzed-options.js [hunk 22]
```diff
@@ -0,0 +1,111 @@
...
+            loc.line,
+            "line",
+            loc,
+          )}:${adjust(adjust, loc.column, "column", loc)})`;
+        }),
+      )
+    : error;
+
+export default function toFuzzedOptions(options) {
+  if (TEST_FUZZ === "false") return [[false, options]];
```
Comment (JLHwung) [inline packages/babel-parser/test/helpers/to-fuzzed-options.js:49]: Can we make `TEST_FUZZ` opt-in? As long as CI will run fuzz testing, we don't have to run them in local environments.

Comment (tolmasky) [reply packages/babel-parser/test/helpers/to-fuzzed-options.js:49]: I'm fine with that. I think I might want to push back on that later when (hopefully) more of the options are tested through fuzz testing (like the errorRecovery I gave above), especially since the entirety of the testing time is dominated by file-reads vs. actual "testing". However, as this only tests the line number stuff right now, I'm fine with reversing this and turning it on in CI.

File: packages/babel-parser/test/helpers/run-fixture-tests.js (hunk 18 above)
Comment (tolmasky) [reply packages/babel-parser/test/helpers/run-fixture-tests.js:128]: Sure thing, just want to confirm that this is new behavior that we want, since I was following the existing code (which doesn't add a newline):

1. https://github.com/babel/babel/blob/main/packages/babel-parser/test/helpers/runFixtureTests.js#L68
2. https://github.com/babel/babel/blob/main/packages/babel-parser/test/helpers/runFixtureTests.js#L138

Super easy and happy to add, just don't want to break anything if something else is expecting it not to have newlines.

Comment (nicolo-ribaudo) [reply packages/babel-parser/test/helpers/run-fixture-tests.js:133]: Oh yes please, even if it means changing every fixture. It's super annoying that when I manually touch an `option.json` then my editor adds a newline which wasn't there.

File: packages/babel-parser/test/helpers/to-fuzzed-options.js (hunk 22 above)
Comment (tolmasky) [reply packages/babel-parser/test/helpers/to-fuzzed-options.js]: Done: https://github.com/babel/babel/pull/14201/commits/b1118e410f39e2f73ade31720e42485de7cf793a

File: packages/babel-parser/test/helpers/run-fixture-tests.js (hunk 21 above)
Comment (tolmasky) [reply packages/babel-parser/test/helpers/run-fixture-tests.js]: Done: https://github.com/babel/babel/pull/14201/commits/dec5379c214d89f0ee6417ef4084cd5efec4964c

File: packages/babel-parser/test/helpers/run-fixture-tests.js (hunk 18 above)
Comment (tolmasky) [reply packages/babel-parser/test/helpers/run-fixture-tests.js]: Done: https://github.com/babel/babel/pull/14201/commits/198ad6bcf2a18fc394de65506ee76d751e58919e

File: packages/babel-parser/test/helpers/run-fixture-tests.js (hunk 19 above)
Comment (tolmasky) [reply packages/babel-parser/test/helpers/run-fixture-tests.js:146]: Done: https://github.com/babel/babel/pull/14201/commits/198ad6bcf2a18fc394de65506ee76d751e58919e

File: packages/babel-parser/test/helpers/run-fixture-tests.js [hunk 23]
```diff
@@ -0,0 +1,201 @@
...
+    return JSONParse(readFileSync(filename, "utf-8"));
+  } catch (error) {
+    return {};
+  }
+}
+
+function rmf(path) {
+  try {
+    unlinkSync(path);
+  } catch (error) {}
```
Comment (JLHwung) [inline packages/babel-parser/test/helpers/run-fixture-tests.js:174]: I think we should throw if `error.code` is not `ENOENT`. We can only handle the case when `options.json` is not found, for other cases better just propagate the errors.

**Review by JLHwung — APPROVED** [review]
Awesome work!

File: packages/babel-parser/test/helpers/run-fixture-tests.js (hunk 23 above)
Comment (tolmasky) [reply packages/babel-parser/test/helpers/run-fixture-tests.js]: Done: https://github.com/babel/babel/pull/14201/commits/91f73c769cb69f4b9c6df7553046c4a65267118d (Also makes the error message when you expect an error but don't get an error slightly nicer).

**Comment by tolmasky** [comment]
Thanks!

**Review by existentialism — APPROVED** [review]
This is awesome!
